Portal/src/Datahub.Infrastructure/Services/Achievements/PortalUserTelemetryService.cs

```diff
@@ -34,6 +39,34 @@ public PortalUserTelemetryService(
 
     public event EventHandler<AchievementsEarnedEventArgs> OnAchievementsEarned;
 
+    public async Task FindUserLastLogin()
+    {
+        await using var ctx = await contextFactory.CreateDbContextAsync();
+
+        var portalUser = await userInformationService.GetCurrentPortalUserWithAchievementsAsync();
+        // check the user exists
+        if (portalUser is null)
+        {
+            logger.LogWarning("Getting user's last login without a Portal User.");
+            return;
+        }
+        var userLogins = ctx.TelemetryEvents.Where(t => t.PortalUserId == portalUser.Id && t.EventName == TelemetryEvents.UserLogin).ToList();
+        if (userLogins.Count > 0)
+        {
+            lastLogin = userLogins.ElementAt(userLogins.Count - 1).EventDate.ToLocalTime();
+        }
+        else
+        {
+            lastLogin = new DateTime(0001, 01, 01);
+        }
```

Comment:
There is a small issue here, this will currently show the _most recent login time_ (since `userLogins.Count -1` is the last element in the list), but what we want is the _previous one_, i.e. `userLogins.Count-2`. The reason why we need to add this is so people can see if the previous login (prior to the most recent one) was done by themselves (and not by a malicious other person).

I would also suggest adding a bit more logic to ensure this never crashes. For example, what I mentioned above requires 2 elements in the list, so I would make sure the count is greater than 1, and otherwise, instead of returning `DateTime(0001, 01, 01)`, you should return `null`: this is pretty important, so we never want to return a made up or incorrect value.

The last part would be to have some logic on the UI if the returned value is `null`

Reply:
Just pushed those changes. The only issue still is with the way we're handling adding a login telemetry event. Currently it's processed in the DatahubAchievementProvider.razor file, which means that when we swap between languages, we're adding another login event. Obviously this isn't what we'd want, but I'm not confident enough with how the login is handled to know where to move `await _telemetryService.LogTelemetryEvent(TelemetryEvents.UserLogin);`